File: CalenderView/app/src/main/java/com/non_name_hero/calenderview/utils/PigLeadUtils.java
package com.non_name_hero.calenderview.utils;

import com.non_name_hero.calenderview.data.CalendarData;
import com.non_name_hero.calenderview.data.Schedule;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class PigLeadUtils {
    //年月日のフォーマット
    public static SimpleDateFormat formatYYYYMMDD = new SimpleDateFormat("yyyy.MM.dd", Locale.US);
    public static SimpleDateFormat yearFormat = new SimpleDateFormat("yyyy", Locale.US);
    public static SimpleDateFormat monthFormat = new SimpleDateFormat("MM", Locale.US);
    public static SimpleDateFormat dayFormat = new SimpleDateFormat("dd", Locale.US);
    public static SimpleDateFormat formatD = new SimpleDateFormat("d", Locale.US);

    /**
     * List<Schedule> から Map<String, List<Schedule>>へ変換する
     * @param schedules
     * @return scheduleMap -　Map<String, List<Schedule>>
     */
    public static Map<String, List<Schedule>> getScheduleMapBySchedules(List<Schedule> schedules){
        Map<String, List<Schedule>> scheduleMap= new HashMap<>();
        List<Schedule> scheduleLists;
        for (Schedule schedule : schedules) {
            String date = formatYYYYMMDD.format(schedule.getStartAtDatetime());
            if (scheduleMap.containsKey(date)) {
                scheduleLists = scheduleMap.get(date);
                scheduleLists.add(schedule);
                scheduleLists = new ArrayList<>(new LinkedHashSet<>(scheduleLists));

            } else {
                scheduleLists = new ArrayList<Schedule>();
                scheduleLists.add(schedule);
            }
            scheduleMap.put(date, scheduleLists);
        }
        return scheduleMap;
    }

    /**
     * List<CalendarData> から Map<String, List<CalendarData>>へ変換する
     * @param calendarDataList
     * @return scheduleMap -　Map<String, List<Calendardata>>
     */
    public static Map<String, List<CalendarData>> getCalendarDataMapByCalendarDataList(List<CalendarData> calendarDataList){
        Map<String, List<CalendarData>> calendarDataMap= new HashMap<>();
        List<CalendarData> calendarDataLists;
        for (CalendarData data : calendarDataList) {
            String date = formatYYYYMMDD.format(data.scheduleStartAtDatetime);
            if (calendarDataMap.containsKey(date)) {
                calendarDataLists = calendarDataMap.get(date);
                calendarDataLists.add(data);
                calendarDataLists = new ArrayList<>(new LinkedHashSet<>(calendarDataLists));
            } else {
                calendarDataLists = new ArrayList<CalendarData>();
                calendarDataLists.add(data);
            }
            calendarDataMap.put(date, calendarDataLists);
        }
        return calendarDataMap;
    }

}

File: CalenderView/app/src/main/java/com/non_name_hero/calenderview/utils/PigLeadUtils.kt
package com.non_name_hero.calenderview.utils

import com.non_name_hero.calenderview.data.BalanceData
import com.non_name_hero.calenderview.data.CalendarData
import com.non_name_hero.calenderview.data.Schedule
import java.text.SimpleDateFormat
import java.util.*

object PigLeadUtils {
    //年月日のフォーマット
    var formatYYYYMMDD = SimpleDateFormat("yyyy.MM.dd", Locale.US)
    var yearFormat = SimpleDateFormat("yyyy", Locale.US)
    var monthFormat = SimpleDateFormat("MM", Locale.US)
    var dayFormat = SimpleDateFormat("dd", Locale.US)
    var formatD = SimpleDateFormat("d", Locale.US)

    /**
     * List<Schedule> から Map<String></String>, List<Schedule>>へ変換する
     * @param schedules
     * @return scheduleMap -　Map<String></String>, List<Schedule>>
    </Schedule></Schedule></Schedule> */
    fun getScheduleMapBySchedules(schedules: List<Schedule>): Map<String, MutableList<Schedule>> {
        val scheduleMap: MutableMap<String, MutableList<Schedule>> = HashMap()
        var scheduleLists: MutableList<Schedule>
        for (schedule in schedules) {
            val date = formatYYYYMMDD.format(schedule.startAtDatetime)
            if (scheduleMap.containsKey(date)) {
                scheduleLists = scheduleMap[date]!!
                scheduleLists.add(schedule)
                scheduleLists = ArrayList(LinkedHashSet(scheduleLists))
            } else {
                scheduleLists = mutableListOf(schedule)
            }
            scheduleMap[date] = scheduleLists
        }
        return scheduleMap
    }

    /**
     * List<CalendarData> から Map<String>, List<CalendarData>>へ変換する
     * @param calendarDataList
     * @return scheduleMap<Calendardata>
     **/
    fun getCalendarDataMapByCalendarDataList(calendarDataList: List<CalendarData>): Map<String, MutableList<CalendarData>> {
        val calendarDataMap: MutableMap<String, MutableList<CalendarData>> = HashMap()
        var calendarDataLists: MutableList<CalendarData>
        for (data in calendarDataList) {
            // 日付キーを生成
            val date = formatYYYYMMDD.format(data.scheduleStartAtDatetime)
            if (calendarDataMap.containsKey(date)) {
                //calendarDataMap[date]が非nullであることは保証される
                calendarDataLists = calendarDataMap[date]!!
                calendarDataLists.add(data)
                // 重複するデータを削除してArrayListを再構成する
                calendarDataLists = ArrayList(LinkedHashSet(calendarDataLists))
            } else {
                calendarDataLists = mutableListOf(data)
            }
            // 日付をキーとして値にデータのリストを格納
            calendarDataMap[date] = calendarDataLists
        }
        return calendarDataMap
    }

    fun getBalanceCalendarDataMapByBalanceDataList(balanceDataList: List<BalanceData>): Map<String, MutableList<BalanceData>> {
        val calendarBalanceDataMap: MutableMap<String, MutableList<BalanceData>> = HashMap()
        var balanceDataLists: MutableList<BalanceData>
        // 日付に家計簿データをマッピング
        for (data in balanceDataList) {
            // 日付キーを生成
            val date = data.timestamp
            if (calendarBalanceDataMap.containsKey(date)) {
                // すでにその日にちのデータがある場合は値を取得した上で追加
                balanceDataLists = calendarBalanceDataMap[date]!!
                // 同じbalanceType(収支タイプ)が同じ要素がなければ要素を追加
                balanceDataLists.add(data)
            } else {
                // すでにその日にちのデータがない場合は新規にArrayList作成
                balanceDataLists = mutableListOf(data)
            }
            // 日付をキーとして値にデータのリストを格納
            calendarBalanceDataMap[date] = balanceDataLists
        }
        return calendarBalanceDataMap
    }
}